src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+  } else {
+    // Input is a scalar value not in this loop
+    assert(!in_body(in), "Node shouldn't be in this loop");
+    if (VectorNode::is_roundopD(node) && idx == 2) {
+      // 1) Just return the scalar input
+      return in;
+    } else {
+      // 2) Need replicate the scalar input
+      Node* vrep = nullptr;
+      if (VectorNode::is_shift(node) && idx == 2) {
+        // 2.1) Input is the 2nd (shift count) of left/right shift
+        assert(is_integral_type(bt), "Shift operation should work on integers");
+        Node* mask_con = _igvn->intcon((bt == T_LONG) ?
+                                       (BitsPerLong - 1) : (BitsPerInt - 1));
+        Node* mask_op = new AndINode(in, mask_con);
+        _igvn->register_new_node_with_optimizer(mask_op);
+        vrep = VectorNode::shift_count(node->Opcode(), mask_op, vlen, bt);
+      } else if (VectorNode::is_scalar_rotate(node) && idx == 2) {
+        // 2.2) Input is the 2nd (rotate shift count) of rotate shift
+        Node* conv = in;
+        if (bt == T_LONG) {
+          conv = new ConvI2LNode(in);
+          _igvn->register_new_node_with_optimizer(conv);
+        }
+        vrep = VectorNode::scalar2vector(conv, vlen, Type::get_const_basic_type(bt));
+      } else {
+        // 2.3) Other general scalar inputs
+        const Type* type = Type::get_const_basic_type(get_signed_subword_bt(bt));
+        vrep = VectorNode::scalar2vector(in, vlen, type);
+      }
+      _igvn->register_new_node_with_optimizer(vrep);
+      VectorNode::trace_new_vector(vrep, "VectorMasked");
+      return vrep;
+    }
+  }
+}
+
+// Replace scalar nodes in the loop by vector nodes from top to bottom and
+// return the node map of scalar to vector replacement
+Node_List* VectorMaskedLoop::replace_scalar_ops(Node* mask) {
+  // Create a node map of scalar to vector replacement
+  int n_nodes = _body_nodes.length();
+  Node_List* s2v_map = new Node_List(_arena, n_nodes);
+
+  // Replace each node with valid element basic type set
+  for (int idx = 0; idx < n_nodes; idx++) {
+    Node* snode = _body_nodes.at(idx);
+    if (has_valid_elem_bt(snode)) {
+      Node* vnode;
+      int opc = snode->Opcode();
+      BasicType bt = elem_bt(snode);
+      int vlen = Matcher::max_vector_size(bt);
+      if (snode->is_Mem()) {
+        Node* ctrl = snode->in(MemNode::Control);
+        Node* mem = snode->in(MemNode::Memory);
+        Node* addr = snode->in(MemNode::Address);
+        const TypePtr* at = snode->as_Mem()->adr_type();
+        const TypeVect* vt = TypeVect::make(Type::get_const_basic_type(bt), vlen);
+        if (snode->is_Load()) {
+          vnode = new LoadVectorMaskedNode(ctrl, mem, addr, at, vt, mask);
+        } else {
+          assert(snode->is_Store(), "Unexpected memory op");
+          Node* val = get_vector_input(snode, MemNode::ValueIn);
+          vnode = new StoreVectorMaskedNode(ctrl, mem, addr, val, at, mask);
+        }
+      } else if (VectorNode::is_convert_opcode(opc)) {
```

Comment:
Ok, this does work for same size conversions:
`./java -Xcomp -XX:-TieredCompilation -XX:+TraceNewVectors -XX:+TraceLoopOpts -XX:+UnlockExperimentalVMOptions -XX:+UseMaskedLoop -XX:+TraceMaskedLoop -XX:CompileCommand=compileonly,Test::test0 -XX:+TraceSuperWord Test.java`
```
public class Test {
    static int RANGE = 1024;

    public static void main(String[] strArr) {
        double a[] = new double[RANGE];
        long b[] = new long[RANGE];
        test0(a, b);
    }

    static void test0(double[] a, long[] b) {
        for (int i = 0; i < RANGE; i++) {
            b[i] = (long)a[i];
        }
    }
}
```
Good to see some conversion is possible. But if I replace double with float, I get `Vector element size does not match`. Can that limitation be lifted?

Reply:
We tried to do that but found some obstacles (perhaps you are aware of).

> If you started implementing type conversion for different size types, you'd have to extract_lo/hi or pack the vectors. That would be an invasive change to the current implementation.

As you said in your overall feedback, conversions between types of different data sizes requires vector pack/unpack which has conflict with existing semantics of current C2 type conversion nodes. We are still considering how to do it. It would be good if you have better suggestions or can help us with this.